sql-plugin/src/main/scala/com/nvidia/spark/rapids/SpillFramework.scala

```diff
@@ -0,0 +1,1606 @@
+ */
+trait StoreHandle extends AutoCloseable {
+  /**
+   * Approximate size of this handle, used in two scenarios:
+   * - Used from the host store to figure out how much host memory total it is tracking.
+   * - If sizeInBytes is 0, the object is tracked by the stores so it can be
+   *   removed on shutdown, or by handle.close, but 0-byte handles are not spillable.
+   */
+  val sizeInBytes: Long
+}
+
+trait SpillableHandle extends StoreHandle {
+  /**
+   * Method called to spill this handle. It can be triggered from the spill store,
+   * or directly against the handle.
+   * @return sizeInBytes if spilled, 0 for any other reason (not spillable, closed)
+   */
+  def spill: Long
+
+  /**
+   * Method used to determine whether a handle tracks an object that could be spilled
+   * @note At the level of `SpillableHandle`, the only requirement of spillability
+   *       is that the size of the handle is > 0. `sizeInBytes` is known at construction, and
+   *       is a `val`.
+   * @return true if currently spillable, false otherwise
+   */
+  def spillable: Boolean = sizeInBytes > 0
+}
+
+/**
+ * Spillable handles that can be materialized on the device.
+ * @tparam T an auto closeable subclass. `dev` tracks an instance of this object,
+ *           on the device.
+ */
+trait DeviceSpillableHandle[T <: AutoCloseable] extends SpillableHandle {
+  var dev: Option[T]
+
+  override def spillable: Boolean = synchronized {
+    super.spillable && dev.isDefined
+  }
+}
+
+/**
+ * Spillable handles that can be materialized on the host.
+ * @tparam T an auto closeable subclass. `host` tracks an instance of this object,
+ *           on the host.
+ */
+trait HostSpillableHandle[T <: AutoCloseable] extends SpillableHandle {
+  var host: Option[T]
+
+  override def spillable: Boolean = synchronized {
+    super.spillable && host.isDefined
+  }
+}
+
+object SpillableHostBufferHandle extends Logging {
+  def apply(hmb: HostMemoryBuffer): SpillableHostBufferHandle = {
+    val handle = new SpillableHostBufferHandle(hmb.getLength, host = Some(hmb))
+    SpillFramework.stores.hostStore.track(handle)
+    // do we care if:
+    // handle didn't fit in the store as it is too large.
+    // we made memory for this so we are going to hold our noses and keep going
+    // we could spill `handle` at this point.
+    handle
+  }
+
+  def createHostHandleWithPacker(
+      chunkedPacker: ChunkedPacker): SpillableHostBufferHandle = {
+    val handle = new SpillableHostBufferHandle(chunkedPacker.getTotalContiguousSize)
+    withResource(
+      SpillFramework.stores.hostStore.makeBuilder(handle)) { builder =>
+      SpillFramework.withChunkedPackBounceBuffer { bb =>
+        while (chunkedPacker.hasNext) {
+          withResource(chunkedPacker.next(bb)) { n =>
+            builder.copyNext(n, Cuda.DEFAULT_STREAM)
+            // we are calling chunked packer on `bb` again each time, we need
+            // to synchronize before we ask for the next chunk
```

Comment:
We may want to change the method name to something like `copyNextAsync` so it's clear this is an asynchronous operation.  Otherwise it wasn't immediately clear to me why we need a sync after this.

Reply:
that's a nice catch, I need to make sure this is **not async**, because it currently asks for a bounce buffer in the disk case. So let me fix this.

Reply:
https://github.com/NVIDIA/spark-rapids/pull/11747/commits/b48b801201ea43254b5e3f9550656f6fa7c4adbe